kube/src/api/dynamic.rs

```diff
@@ -1,32 +1,87 @@
-use crate::{
-    api::{metadata::TypeMeta, Resource},
-    Error, Result,
-};
+use crate::api::{metadata::TypeMeta, GroupVersionKind, Resource};
 use k8s_openapi::apimachinery::pkg::apis::meta::v1::{APIResource, ObjectMeta};
-use serde::{Deserialize, Serialize};
 use std::borrow::Cow;
 
-/// Represents a type-erased object kind
-#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq, Hash)]
-pub struct GroupVersionKind {
-    /// API group
-    group: String,
-    /// Version
-    version: String,
-    /// Kind
-    kind: String,
-    /// Concatenation of group and version
-    #[serde(default)]
-    api_version: String,
-    /// Optional plural/resource
-    plural: Option<String>,
+/// Resource scope
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum Scope {
+    /// Objects are global
+    Cluster,
+    /// Each object lives in namespace.
+    Namespaced,
 }
 
-impl GroupVersionKind {
-    /// Creates `GroupVersionKind` from an [`APIResource`].
+/// Operations that are supported on the resource
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct Operations {
+    /// Object can be created
+    pub create: bool,
+    /// Single object can be queried
+    pub get: bool,
+    /// Multiple objects can be queried
+    pub list: bool,
+    /// A watch can be started
+    pub watch: bool,
+    /// A single object can be deleted
+    pub delete: bool,
+    /// Multiple objects can be deleted
+    pub delete_collection: bool,
+    /// Object can be updated
+    pub update: bool,
+    /// Object can be patched
+    pub patch: bool,
+    /// All other verbs
+    pub other: Vec<String>,
+}
+
+impl Operations {
+    /// Returns empty `Operations`
+    pub fn empty() -> Self {
+        Operations {
+            create: false,
+            get: false,
+            list: false,
+            watch: false,
+            delete: false,
+            delete_collection: false,
+            update: false,
+            patch: false,
+            other: Vec::new(),
+        }
+    }
+}
+
+/// Contains information about Kubernetes API resources
+/// which is either required or helpful for working with it.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct ApiResource {
+    /// Resource group, empty for core group.
+    pub group: String,
+    /// group version
+    pub version: String,
+    /// apiVersion of the resource (v1 for core group,
+    /// groupName/groupVersions for other).
+    pub api_version: String,
+    /// Singular PascalCase name of the resource
+    pub kind: String,
+    /// Plural name of the resource
+    pub plural_name: String,
+    /// Scope of the resource
+    pub scope: Scope,
```

Comment:
Alternatively, these three fields can be moved to another type (because they are not used to construct queries), or even deleted.